org.lflang/src/org/lflang/validation/LFValidator.java

```diff
@@ -1199,6 +1206,122 @@ public void checkVarRef(VarRef varRef) {
         }
     }
 
+
+    /**
+     * Specification of the structure of an annotation.
+     */
+    static class AttributeSpec {
```

Comment:
This seems like a good candidate to factor out into its own file.

Reply:
Everything until line 1296 could be moved there.

Reply:
Factoring out seems a bit tricky, since `check()` (L1229) calls `error()`, which is a protected method inherited from `AbstractDeclarativeValidator`. We could move the check functions back to the validator class, but that would sort of invalidate the refactoring because most of the code is still in the validator class.

Reply:
Could pass in a lambda?

Reply:
We can override the error function and just call super, this will make it visible

Reply:
Works like a magic. Thanks for the great advice.